src/futures.rs

```diff
@@ -0,0 +1,73 @@
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::{Arc, Mutex};
+use std::task::{self, Poll, Waker};
+use std::thread;
+
+struct SharedState<T: Send + 'static> {
+    result: Option<T>,
+    waker: Option<Waker>,
+}
+
+pub struct ThreadedFuture<T: Send + 'static> {
+    shared_state: Arc<Mutex<SharedState<T>>>,
+}
+
+impl<T: Send + 'static> ThreadedFuture<T> {
+    pub fn new(f: impl FnOnce() -> T + Send + 'static) -> ThreadedFuture<T> {
+        let shared_state = Arc::new(Mutex::new(SharedState {
+            result: None,
+            waker: None,
+        }));
+
+        // Clone everything to avoid lifetimes
+        let thread_shared_state = shared_state.clone();
+        thread::spawn(move || {
+            let result = f();
+
+            let mut shared_state = thread_shared_state
+                .lock()
+                .expect("ThreadedFuture shared state shouldn't be poisoned");
+
+            shared_state.result = Some(result);
+            if let Some(waker) = shared_state.waker.take() {
+                waker.wake();
+            }
+        });
+
+        ThreadedFuture { shared_state }
+    }
+}
```

Comment:
This is probably the biggest add, here. I'm going to return in a future PR and make use of `async_std` everywhere. Once we are using `async_std`, we can replace this with [`task::spawn_blocking`](https://docs.rs/async-std/1.5.0/async_std/task/fn.spawn_blocking.html), which is essentially the same as what we're doing here.